### PR TITLE
[utils] `js_to_json`: Fix unneccesary escape handling

### PR DESCRIPTION
**Pull request by Grub4K**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

`js_to_json(R'"\^\$\#"')` currently results in invalid json since only specific escapes are getting processed.
This PR provides an alternate implementation with tests against this.
It also tries to clean up the function a bit.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/utils.py [hunk 1]
```diff
@@ -3275,6 +3275,8 @@ def strip_jsonp(code):
 
 def js_to_json(code, vars={}, *, strict=False):
     # vars is a dict of var, val pairs to substitute
+    STRING_QUOTES = '\'\"'
```
Comment (pukkandan) [inline yt_dlp/utils.py:3278]:
```suggestion
    STRING_QUOTES = '\'"'
```

File: yt_dlp/utils.py [hunk 2]
```diff
@@ -3275,6 +3275,8 @@ def strip_jsonp(code):
...
+    STRING_RE = '|'.join(rf'''{q}(?:[^{q}\\]*(?:\\.))*[^{q}\\]*{q}''' for q in STRING_QUOTES)
```
Comment (pukkandan) [inline yt_dlp/utils.py:3279]:
```suggestion
    STRING_RE = '|'.join(rf'{q}(?:[^{q}\\]*(?:\\.))*[^{q}\\]*{q}' for q in STRING_QUOTES)
```
OR
untested, but this should work better
```suggestion
    STRING_RE = '|'.join(rf'{q}(?:(?<!\\)(?:\\\\)*\\{q}|[^{q}])*{q}' for q in STRING_QUOTES)
```

File: yt_dlp/utils.py [hunk 3]
```diff
@@ -3289,28 +3291,34 @@ def fix_kv(m):
...
+
+        if v[0] in STRING_QUOTES:
+            def process_escape(match):
+                JSON_PASSTHROUGH_ESCAPES = R'"\bfnrtu'
+                escape = match.group(1) or match.group(2)
+
+                return (Rf'\{escape}' if escape in JSON_PASSTHROUGH_ESCAPES
+                        else R'\u00' if escape == 'x'
+                        else '' if escape == '\n'
+                        else escape)
```
Comment (pukkandan) [inline yt_dlp/utils.py:3304]: move this outside fix_kv imo

File: yt_dlp/utils.py (hunk 2 above)
Comment (Grub4K) [reply yt_dlp/utils.py:3279]: Actually, whats speaking against `rf'{q}(?:\\.|[^\\{q}])*{q}'`?

File: test/test_utils.py [hunk 4]
```diff
@@ -1100,6 +1100,12 @@ def test_js_to_json_edgecases(self):
         on = js_to_json('[1,//{},\n2]')
         self.assertEqual(json.loads(on), [1, 2])
 
+        on = js_to_json(R'"\^\$\#"')
+        self.assertEqual(json.loads(on), R"^$#", msg="Unneccesary escapes should be stripped")
+
+        on = js_to_json('\'"\\""\'')
+        self.assertEqual(json.loads(on), '"""', msg="Unneccesary quote escape should be escaped")
```
Comment (pukkandan) [inline test/test_utils.py:1107]:
```suggestion
        on = js_to_json(R'"\^\$\#"')
        self.assertEqual(json.loads(on), R'^$#', msg='Unnecessary escapes should be stripped')

        on = js_to_json('\'"\\""\'')
        self.assertEqual(json.loads(on), '"""', msg='Unnecessary quote escape should be escaped')
```

**Pull request merged**